### PR TITLE
[Structural] Dynamic and OSS mixed displacement - volumetric strain elements

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**📝 Description**
In this PR I'm extending the mixed displacement - volumetric strain element to the dynamic regime, both with ASGS and OSGS stabilization. In here only the required variables, elements and unittests are added. I'll add the extra changes required for its usage in a follow up PR. In this regard, I note that the commented patch test will be activated in such follow up PR.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp [hunk 1]
```diff
@@ -1014,6 +1339,69 @@ double SmallDisplacementMixedVolumetricStrainElement::CalculateShearModulus(cons
...
+        noalias(aux) += prod(trans(B_i), Matrix(prod(rThisConstitutiveVariables.D, B_i)));
+    }
+    double det;
+    const double c_1 = 1.0;
+    Matrix tau_1_mat(dim, dim);
+    MathUtils<double>::InvertMatrix(aux, tau_1_mat, det);
+    const Vector tau_1_vect = MathUtils<double>::SymmetricTensorToVector(tau_1_mat);
+    const double m_T_tau_1 = c_1*inner_prod(rVoigtIdAnysotropyMatrixProd, tau_1_vect);
+
+    // Uncomment this to use the stabilisation parameters defined in the paper
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp:1371]: Does not work?

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp [hunk 2]
```diff
@@ -1143,32 +1597,134 @@ void SmallDisplacementMixedVolumetricStrainElement::CalculateOnIntegrationPoints
...
             "nodal_non_historical"   : [],
-            "entity"                 : []
+            "entity"                 : ["DISPLACEMENT_SUBSCALE", "VOLUMETRIC_STRAIN_SUBSCALE"]
         },
         "required_variables"         : ["DISPLACEMENT","VOLUMETRIC_STRAIN"],
         "required_dofs"              : [],
         "flags_used"                 : [],
         "compatible_geometries"      : ["Triangle2D3", "Quadrilateral2D4", "Tetrahedra3D4","Hexahedra3D8"],
-        "element_integrates_in_time" : true,
+        "element_integrates_in_time" : false,
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp:1719]: It is not true now and false then?

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp [hunk 3]
```diff
@@ -1194,6 +1750,8 @@ void SmallDisplacementMixedVolumetricStrainElement::save(Serializer& rSerializer
     rSerializer.save("ConstitutiveLawVector", mConstitutiveLawVector);
     rSerializer.save("AnisotropyTensor", mAnisotropyTensor);
     rSerializer.save("InverseAnisotropyTensor", mInverseAnisotropyTensor);
+    // rSerializer.save("DisplacementSubscale1", mDisplacementSubscale1);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: Does not compile?

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_small_displacement_mixed_volumetric_strain_oss_element.cpp [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    const std::vector<double> expected_LHS_row_0({778846.153846, 9615.38461538, -327036.58062, -394230.769231, -384615.384615, -307578.803995, -384615.384615, 375000, -317307.692308});
+    const std::vector<double> expected_mass_row_0({83.3333333333, 0, 0, 41.6666666667, 0, 0, 41.6666666667, 0, 0});
+    const std::vector<double> expected_mass_row_2({0, 0, 0, 0, 0, 0, 0, 0, 0});
+    KRATOS_EXPECT_VECTOR_RELATIVE_NEAR(RHS, expected_RHS, tolerance)
+    KRATOS_EXPECT_VECTOR_RELATIVE_NEAR(row(LHS,0), expected_LHS_row_0, tolerance)
+    KRATOS_EXPECT_VECTOR_RELATIVE_NEAR(row(MassMatrix,0), expected_mass_row_0, tolerance)
+    KRATOS_EXPECT_VECTOR_RELATIVE_NEAR(row(MassMatrix,2), expected_mass_row_2, tolerance)
+}
+
+// KRATOS_TEST_CASE_IN_SUITE(SmallDisplacementMixedVolumetricStrainOssElementZienkiewiczPatch, KratosStructuralMechanicsFastSuite)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/tests/cpp_tests/test_small_displacement_mixed_volumetric_strain_oss_element.cpp:175]: ¿?

**Review by loumalouomega — DISMISSED** [review]
Some comments

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp:1371]: What you mean by does not work?

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp (hunk 2 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp:1719]: This means that the time integration is done outside the element (via scheme) so `false` is correct.

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp (hunk 3 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: It should. Most probably I commented it at some point while developing. Thanks for spotting.

File: applications/StructuralMechanicsApplication/tests/cpp_tests/test_small_displacement_mixed_volumetric_strain_oss_element.cpp (hunk 4 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/tests/cpp_tests/test_small_displacement_mixed_volumetric_strain_oss_element.cpp:175]: See the PR description. These will be activated in a follow up PR as require more developments.

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp (hunk 1 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp:1371]: The reason it is commented

Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp:1371]: Ah. To make switching from one definition to the other more handy, there is no special reason.

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**